clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -238,12 +237,20 @@ public boolean rename(Path src, Path dst) throws IOException {
         // https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java#L1505
         LakeFSFileStatus srcStatus;
         srcStatus = getFileStatus(src);
-        if (!srcStatus.isDirectory()) {
-            return renameFile(srcStatus, dst);
+        boolean result;
+        if (srcStatus.isDirectory()) {
+            result = renameDirectory(src, dst);
+        } else {
+            result = renameFile(srcStatus, dst);
+        }
+        if (src.getParent() != dst.getParent()) {
+            deleteUnnecessaryFakeDirectories(dst.getParent());
+            createFakeDirectoryIfNecessary(src.getParent());
```

Comment:
Updated the code to fix the behavior - had to enable the iterator to return directories (markers in this case) when we do a recursive listing. It will move the markers to the destination.

The destination directory now holds a new directory or file, that should have markers (as you said in item 1) we call to delete all the markers from the parent directories.

The source directory is left without the file/directory - we check if the source requires a marker because we possibly left it empty.

Currently, there is just one scan up for clean markers assuming we placed something under that path and all these markers are irrelevant.
Which operations in parallel or scenario are you worried about? or just from design pov you prefer one function to do the delete/create?